File: backend/src/main/scala/domains/blog/api/AcceptBlogProblemSubmission.scala
package domains.blog.api

import cats.effect.IO
import domains.auth.api.AuthenticatedApi
import domains.auth.objects.internal.AuthenticatedUser
import domains.blog.table.blog.BlogProblemLinkMutationTable
import io.circe.Encoder
import org.http4s.{Method, Request, Status}
import shared.api.{ApiMessages, ApiPath, HttpApiError, PathParams}
import shared.objects.response.SuccessResponse

import java.sql.Connection

object AcceptBlogProblemSubmission extends AuthenticatedApi[BlogProblemLinkInput, SuccessResponse]:

  override val method: Method = Method.POST
  override val path: ApiPath = ApiPath("/api/problems/:problemSlug/blog-submissions/:blogId/accept")
  override val successStatus: Status = Status.Ok
  override protected val outputEncoder: Encoder[SuccessResponse] = summon[Encoder[SuccessResponse]]

  override def decode(request: Request[IO], pathParams: PathParams): IO[BlogProblemLinkInput] =
    val _ = request
    BlogProblemLinkInput.fromPathParams(pathParams)

  override def plan(connection: Connection, actor: AuthenticatedUser, input: BlogProblemLinkInput): IO[SuccessResponse] =
    for
      _ <- ProblemBlogAccess.requireProblemCatalogManager(actor)
      accepted <- BlogProblemLinkMutationTable.acceptProblem(connection, input.problemSlug, input.blogId, actor.username)
      _ <- HttpApiError.ensure(accepted, HttpApiError.notFound(ApiMessages.pendingProblemBlogSubmissionNotFound))
    yield SuccessResponse.fromApiMessage(ApiMessages.problemBlogSubmissionAccepted)

File: frontend/src/apis/blog/AcceptBlogProblemSubmission.ts
import type { APIWithSessionMessage } from '@/system/api/api-message'
import type { BlogId } from '@/objects/blog/BlogId'
import { blogIdValue } from '@/objects/blog/BlogId'
import type { ProblemSlug } from '@/objects/problem/ProblemSlug'
import { problemSlugValue } from '@/objects/problem/ProblemSlug'
import type { SuccessResponse } from '@/objects/shared/transport/SuccessResponse'

/** 接受博客投稿关联到题目；输入题目 slug 和博客 ID，输出通用成功响应。 */
export class AcceptBlogProblemSubmission implements APIWithSessionMessage<SuccessResponse> {
  declare readonly responseType?: SuccessResponse
  readonly method = 'POST'
  readonly apiPath: string

  constructor(problemSlug: ProblemSlug, blogId: BlogId) {
    this.apiPath = `problems/${problemSlugValue(problemSlug)}/blog-submissions/${blogIdValue(blogId)}/accept`
  }

  body(): undefined {
    return undefined
  }
}
